datafusion/core/src/datasource/function.rs

```diff
@@ -0,0 +1,60 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! A table that uses a function to generate data
+
+use std::sync::Arc;
+
+use super::{streaming::StreamingTable, TableProvider};
+use arrow::datatypes::SchemaRef;
+use datafusion_common::Result;
+use datafusion_physical_plan::streaming::PartitionStream;
+
+/// Table Function implementation
+pub type TableFunctionImplementation = Arc<
+    dyn Fn(&[String]) -> Result<(SchemaRef, Vec<Arc<dyn PartitionStream>>)> + Send + Sync,
+>;
+
+/// A table that uses a function to generate data
+pub struct TableFunction {
+    /// Name of the table function
+    name: String,
+    /// Function implementation
+    fun: TableFunctionImplementation,
+}
+
+impl TableFunction {
+    /// Create a new table function
+    pub fn new(name: String, fun: TableFunctionImplementation) -> Self {
+        Self { name, fun }
+    }
+
+    /// Get the name of the table function
+    pub fn name(&self) -> String {
+        self.name.clone()
+    }
+
+    /// Get the function implementation and generate a table
+    pub fn create_table_provider(
```

Comment:
I was thinking we would allow users to provide their own `TableProvider` directly (rather than a `PartitionStream`) as this permits them to implement predicate / projection / limit pushdown, among other things

I like how you have made a simple version in the example -- maybe you can just move the creation of `StreamingTable::try_new()` into `datafusion-examples/examples/simple_udtf.rs` 🤔 

I am also thinking about how to support TableFunctions that take streams as input: https://github.com/apache/arrow-datafusion/issues/7926#issuecomment-1813692801. Any ideas on this?

Reply:
> I was thinking we would allow users to provide their own TableProvider directly (rather than a PartitionStream) as this permits them to implement predicate / projection / limit pushdown, among other things

 sure! Agree with you, just implementing their `TableProvider` seems more reasonable. I have change the func signature.

> I am also thinking about how to support TableFunctions that take streams as input:

maybe users can implement `PartitionStream` and use inner `StreamTable` to take streams as input? see it in [example](https://github.com/apache/arrow-datafusion/pull/8306/files#diff-97acd33805992e651433d8093b279fb814ec4d0f88261ece537d55f98b5dbf33R111-R140)